### DISCUSSION TITLE
Error compiling on Windows 10, Optix 7.3

### DISCUSSION THREAD

**Discussion by drperpen** [NVlabs/instant-ngp#341] [Q&A] [answered]
Hi,
I get errors of this type, all of them related to _optix_program.vcxproj_:

`C:\Program Files (x86)\Microsoft Visual Studio\2019\Community\VC\Tools\MSVC\14.29.30133\include\xutility(1309): error :
 expected a "(" [E:\dev\instant-ngp\build\optix_program.vcxproj]`

Not sure if this is Visual Studio 2019 related or Optix related.
Thanks!

**Comment by Tom94** [comment]
To be able to help, further information about your compiler versions, cmake version, operating system, GPU(s), as well as full logs are required.

**Comment by drperpen** [comment]
Hi @Tom94 and thank you for your reply and your amazing project, which I believe sets a milestone, if not a turning point, in the field of 3d representation from 2d images. Gratulieren und vielen Danke!

I made it work with no issues on Ubuntu 18.04, but still struggling on Windows 10. My GPU is Nvidia Quadro GTX 5000 (16 Gb) and here is the output of the files build:

`
E:\dev\windows\instant-ngp>cmake . -B build
-- Building for: Visual Studio 16 2019
-- Selecting Windows SDK version 10.0.19041.0 to target Windows 10.0.19042.
-- The C compiler identification is MSVC 19.29.30141.0
-- The CXX compiler identification is MSVC 19.29.30141.0
-- The CUDA compiler identification is NVIDIA 11.3.58
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/VC/Tools/MSVC/14.29.30133/bin/Hostx64/x64/cl.exe - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: C:/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.3/bin/nvcc.exe - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - not found
-- Found Threads: TRUE
-- Using Win32 for window creation
-- Found OpenMP_C: -openmp (found version "2.0")
-- Found OpenMP_CXX: -openmp (found version "2.0")
-- Found OpenMP: TRUE (found version "2.0")
-- OptiX_INSTALL_DIR value: C:\OptiX_SDK_7.3.0
-- Found Python: C:/Users/XXXXXXX/anaconda3/python.exe (found suitable version "3.8.8", minimum required is "3.7") found components: Interpreter Development Development.Module Development.Embed
-- pybind11 v2.7.1
-- Performing Test HAS_MSVC_GL_LTCG
-- Performing Test HAS_MSVC_GL_LTCG - Success
-- Targeting GPU architectures: 75
-- Configuring done
-- Generating done
-- Build files have been written to: E:/dev/windows/instant-ngp/build`

And this is the CMakeError.log:

`Determining if the include file pthread.h exists failed with the following output:
Change Dir: E:/dev/windows/instant-ngp/build/CMakeFiles/CMakeTmp

Run Build Command(s):C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/MSBuild/Current/Bin/MSBuild.exe cmTC_0e265.vcxproj /p:Configuration=Debug /p:Platform=x64 /p:VisualStudioVersion=16.0 /v:m && Microsoft (R) Build Engine version 16.11.2+f32259642 for .NET Framework

Copyright (C) Microsoft Corporation. All rights reserved.



  Microsoft (R) C/C++ Optimizing Compiler Version 19.29.30141 for x64

  Copyright (C) Microsoft Corporation.  All rights reserved.

  CheckIncludeFile.c

  cl /c /Zi /W1 /WX- /diagnostics:column /Od /Ob0 /D WIN32 /D _WINDOWS /D "CMAKE_INTDIR=\"Debug\"" /D _MBCS /Gm- /RTC1 /MDd /GS /fp:precise /Zc:wchar_t /Zc:forScope /Zc:inline /Fo"cmTC_0e265.dir\Debug\\" /Fd"cmTC_0e265.dir\Debug\vc142.pdb" /external:W1 /Gd /TC /errorReport:queue "E:\dev\windows\instant-ngp\build\CMakeFiles\CMakeTmp\CheckIncludeFile.c"

E:\dev\windows\instant-ngp\build\CMakeFiles\CMakeTmp\CheckIncludeFile.c(1,10): fatal error C1083: Cannot open include file: 'pthread.h': No such file or directory [E:\dev\windows\instant-ngp\build\CMakeFiles\CMakeTmp\cmTC_0e265.vcxproj]`

**Comment by leventt** [comment]
@drperpen try clearing the cmake cache or removing/renaming the build folder and run the cmake commands to configure and build in a developer command line, if you aren't already using that. also pass in visual studio as the generated target solution by passing this to cmake while configuring:
```
cmake -G "Visual Studio 16 2019" ...
```

CUDA extensions for building with Visual Studio are installed part of the CUDA installation if it finds the Visual Studio toolchain and other toolchains like Ninja may default to some MSVC version that didn't get that CUDA installer treatment.

I have a feeling, it may be due cmake generating for another toolchain but more likely, the build environment not being set in command line. Default cmd doesn't set the env vars for building... You will have to search on windows for something like this instead of cmd `x64 Native Tools Command Prompt for VS 2019` It is only installed with Visual Studio along with command line toolchain... sometimes you will need to specify this to be installed in the installer or sometimes you will need to create a blank project in visual studio community edition to skip past visual studio pro free trial trap.

**Comment by Tungway1990** [comment, marked as answer]
Actually, it is not related to Optix.

I experience similar error with MS VS 19, CUDA 11.3 and Optix 7.4

After I upgrade CUDA to 11.6, the errors are gone

**Reply by drperpen** [reply]
Thanks, using CUDA 11.6 and Cmake GUI made it.

**Reply by jhgvc** [reply]
我操你妈，搁那瞎指挥，一点屁用没有
